sympy/solvers/bivariate.py

```diff
@@ -176,22 +185,85 @@ def _lambert(eq, x):
     p, den = den.as_coeff_Mul()
     e = exp(num/den)
     t = Dummy('t')
-    args = [d/(a*b)*t for t in roots(t**p - e, t).keys()]
-
     # calculating solutions from args
-    for arg in args:
+    if p == 1:
+        t = e
+        args = [d/(a*b)*t]
+    elif domain is S.Reals:
+        ind_ls = [d/(a*b)*t for t in roots(t**p - e, t).keys()]
+        args = []
+        j = -1
+        for i in ind_ls:
+            j += 1
+            if not isinstance(i,int):
+                if not i.has(I):
+                    args.append(ind_ls[j])
+            elif isinstance(i,int):
+                args.append(ind_ls[j])
+    else:
+        args = [d/(a*b)*t for t in roots(t**p - e, t).keys()]
```

Comment:
There are many places in this PR where the code is unnecessarily convoluted. For example I think all of the code above can just be replaced with:
```python
root_list = list(roots(t**p - e, t))
if domain is S.Reals:
    root_list = [r for r in root_list if not r.has(I)]
args = [r*d/(a*b) for r in root_list]
```